src/Pluggable.groovy

```diff
@@ -0,0 +1,33 @@
+trait Pluggable<T> implements Resettable {
+    private static plugins = []
+    private appliedPlugins = []
+
+    public applyPlugins() {
+        def remainingPlugins = plugins - appliedPlugins
+
+        for (T plugin in remainingPlugins) {
+            plugin.apply(this)
+            appliedPlugins << plugin
+        }
+    }
+
+    public static void addPlugin(T plugin) {
+        plugins << plugin
+    }
+
+    public static void setPlugins(plugins) {
+        this.plugins = plugins
+    }
+
+    public static getPlugins() {
+        return plugins
+    }
+
+    public static void resetPlugins(defaultPlugins = []) {
+        this.plugins = defaultPlugins.clone()
+    }
+
+    public static void reset() {
+        this.resetPlugins()
+    }
```

Comment:
Just to confirm the expected behavior here: 
1. If I implement Pluggable, I'll automatically get a `reset()` that resets my internal plugins
2. If I have additional things that need to be reset, I can override `reset()`, lose access to `Pluggable.reset()` (by nature of the way Groovy Trait static methods work), but can still continue to reset my internal plugins by calling `resetPlugins()` which will still be accessible.

Is that correct?

Reply:
Actually, I think you answer this clearly, in your change to `TerraformApplyCommand`, and `TerraformFormatCommand`.

1.  `TerraformApplyCommand` gets `reset()` "for free" from Pluggable
2. `TerraformFormatCommand` has additional things that need to be reset, and can override and do so, while maintaining plugin reset through `resetPlugins()`

Reply:
That is correct.